src/lib/comms/include/sol-http-client.h

```diff
@@ -55,12 +55,23 @@ extern "C" {
 
 struct sol_http_client_connection;
 
+struct sol_http_request_interface {
+    size_t (*write_cb)(void *userdata, const struct sol_http_client_connection *connection, struct sol_buffer *buffer);
+    size_t (*read_cb)(void *userdata, const struct sol_http_client_connection *connection, void *buffer, size_t buffer_size);
+    void (*response_cb)(void *userdata, const struct sol_http_client_connection *connection, struct sol_http_response *response);
+};
+
 struct sol_http_client_connection *sol_http_client_request(enum sol_http_method method,
     const char *base_uri, const struct sol_http_params *params,
     void (*cb)(void *data, const struct sol_http_client_connection *connection,
     struct sol_http_response *response),
     const void *data) SOL_ATTR_NONNULL(2, 4) SOL_ATTR_WARN_UNUSED_RESULT;
 
+struct sol_http_client_connection *sol_http_client_request_with_interface(enum sol_http_method method,
```

Comment:
document both functions (actually all in this header), make sure you document the expected behavior, like if `interface` can be null, if `interface->response_cb` can be null, if `interface->write_cb` can be null... etc.

Cross link using `@see XXX` to things like `sol_http_client_connection_cancel()` and the "simpler"  version `sol_http_client_request()` (as well as the other way around).